File: [100] Same Tree/100.same-tree.cpp
/*
 * @lc app=leetcode id=100 lang=cpp
 *
 * [100] Same Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        // DFS (with stack)
        if (p == nullptr && q == nullptr) return true;
        if (p == nullptr || q == nullptr) return false;
        stack<TreeNode*> st;
        st.push(q);
        st.push(p);
        while (!st.empty()) {
            p = st.top();
            st.pop();
            q = st.top();
            st.pop();
            if (p == nullptr && q == nullptr) continue;
            if (p == nullptr || q == nullptr || p->val != q->val) return false;
            st.push(q->right);
            st.push(p->right);
            st.push(q->left);
            st.push(p->left);
        }
        return true;
    }
};
// @lc code=end

File: [100] Same Tree/100.same-tree.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode id=100 lang=java
 *
 * [100] Same Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        // BFS (with queue)
        if (p == null && q == null) return true;
        if (p == null || q == null) return false;
        Queue<TreeNode> que_p = new LinkedList<>();
        Queue<TreeNode> que_q = new LinkedList<>();
        que_p.offer(p);
        que_q.offer(q);
        while (!que_p.isEmpty()) {
            p = que_p.poll();
            q = que_q.poll();
            if (p == null && q == null) continue;
            if (p == null || q == null || p.val != q.val) return false;
            que_p.offer(p.left);
            que_q.offer(q.left);
            que_p.offer(p.right);
            que_q.offer(q.right);
        }
        return true;
    }
}
// @lc code=end

